pypeit/edgetrace.py

```diff
@@ -3572,6 +3584,43 @@ def peak_refine(self, rebuild_pca=False, debug=False):
         if ntrace < self.ntrace:
             msgs.warn('Found fewer traces using peak finding than originally available.  '
                       'May want to reset peak threshold.')
+            
+        if self.par['trace_rms_tol'] is not None:
+            # Get the PCA reference row.  The PCA *must* have been defined to get
+            # this far (see pcatype test at the beginning of the function).
+            reference_row = self.left_pca.reference_row if self.par['left_right_pca'] \
+                                else self.pca.reference_row
+            # Match the new trace positions to the input ones
+            gpm = np.logical_not(self.bitmask.flagged(self.edge_msk[reference_row]))
+            # TODO: Include a tolerance here?  Needs testing with more datasets.
+            peak_indx = slitdesign_matching.match_positions_1D(
+                            self.edge_fit[reference_row][gpm],
+                            fit[reference_row])
+
+            # Determine the RMS difference between the input and output traces.
+            # This allows us to compare traces that had already been identified
+            # to their new measurements resulting from peak_trace, and remove
+            # them if they are too discrepant from their original form.  This is
+            # largely meant to find and remove poorly constrained traces, where
+            # the polynomial fit goes wonky.
+            diff = fit - self.edge_fit.T[gpm][peak_indx].T
+            rms = np.sqrt(np.mean((diff - np.mean(diff, axis=0)[None,:])**2, axis=0))
+
+            # TODO: Add a report to the screen or a QA plot?
```

Comment:
I think it might be good to print the rms of each order to screen.

Reply:
Done!  Here's an example:
```
[INFO]    :: edgetrace.py 3610 peak_refine() - ------------------------------
[INFO]    :: edgetrace.py 3611 peak_refine() - Matched spatial locations and RMS difference along spectral direction
[INFO]    :: edgetrace.py 3612 peak_refine() -       OLD      NEW      RMS
[INFO]    :: edgetrace.py 3613 peak_refine() -  -------- -------- --------
[INFO]    :: edgetrace.py 3617 peak_refine() -      13.6     18.6   25.434
[INFO]    :: edgetrace.py 3617 peak_refine() -      37.2     37.2    0.015
[INFO]    :: edgetrace.py 3617 peak_refine() -      74.2     74.2    0.041
[INFO]    :: edgetrace.py 3617 peak_refine() -     112.0    112.0    0.062
[INFO]    :: edgetrace.py 3617 peak_refine() -     150.6    150.6    0.091
[INFO]    :: edgetrace.py 3617 peak_refine() -     232.4    232.4    0.088
[INFO]    :: edgetrace.py 3617 peak_refine() -     274.1    274.1    0.078
[INFO]    :: edgetrace.py 3617 peak_refine() -     317.4    317.3    0.035
[INFO]    :: edgetrace.py 3617 peak_refine() -     362.2    362.2    0.075
[INFO]    :: edgetrace.py 3617 peak_refine() -     408.5    408.4    0.103
[INFO]    :: edgetrace.py 3617 peak_refine() -     456.2    456.2    0.132
[INFO]    :: edgetrace.py 3617 peak_refine() -     505.4    505.4    0.160
[INFO]    :: edgetrace.py 3617 peak_refine() -     556.3    556.3    0.193
[INFO]    :: edgetrace.py 3617 peak_refine() -     609.0    609.0    0.223
[INFO]    :: edgetrace.py 3617 peak_refine() -     663.5    663.4    0.223
[INFO]    :: edgetrace.py 3617 peak_refine() -     720.1    720.1    0.225
[INFO]    :: edgetrace.py 3617 peak_refine() -     778.6    778.6    0.175
[INFO]    :: edgetrace.py 3617 peak_refine() -     839.3    839.3    0.124
[INFO]    :: edgetrace.py 3617 peak_refine() -     902.4    902.4    0.076
[INFO]    :: edgetrace.py 3617 peak_refine() -     967.9    967.9    0.023
[INFO]    :: edgetrace.py 3617 peak_refine() -    1036.1   1036.1    0.051
[INFO]    :: edgetrace.py 3617 peak_refine() -    1107.2   1107.2    0.126
[INFO]    :: edgetrace.py 3617 peak_refine() -    1181.5   1181.4    0.208
[INFO]    :: edgetrace.py 3617 peak_refine() -    1259.1   1259.0    0.299
[INFO]    :: edgetrace.py 3617 peak_refine() -    1340.3   1340.3    0.319
[INFO]    :: edgetrace.py 3617 peak_refine() -    1426.3   1426.3    0.483
[INFO]    :: edgetrace.py 3617 peak_refine() -    1515.8   1515.7    0.367
[INFO]    :: edgetrace.py 3617 peak_refine() -    1609.9   1609.9    0.237
[INFO]    :: edgetrace.py 3617 peak_refine() -    1709.3   1709.3    0.090
[INFO]    :: edgetrace.py 3617 peak_refine() -    1814.3   1814.3    0.088
[INFO]    :: edgetrace.py 3617 peak_refine() -    1925.8   1925.8    0.277
[INFO]    :: edgetrace.py 3617 peak_refine() -      42.3     42.3    0.084
[INFO]    :: edgetrace.py 3617 peak_refine() -      78.1     78.1    0.075
[INFO]    :: edgetrace.py 3617 peak_refine() -     114.8    114.8    0.047
[INFO]    :: edgetrace.py 3617 peak_refine() -     152.8    152.8    0.020
[INFO]    :: edgetrace.py 3617 peak_refine() -     230.4    230.4    0.132
[INFO]    :: edgetrace.py 3617 peak_refine() -     272.0    272.1    0.120
[INFO]    :: edgetrace.py 3617 peak_refine() -     314.5    314.5    0.047
[INFO]    :: edgetrace.py 3617 peak_refine() -     357.9    357.9    0.048
[INFO]    :: edgetrace.py 3617 peak_refine() -     402.6    402.7    0.071
[INFO]    :: edgetrace.py 3617 peak_refine() -     448.8    448.8    0.096
[INFO]    :: edgetrace.py 3617 peak_refine() -     496.4    496.4    0.120
[INFO]    :: edgetrace.py 3617 peak_refine() -     545.6    545.6    0.146
[INFO]    :: edgetrace.py 3617 peak_refine() -     596.4    596.4    0.174
[INFO]    :: edgetrace.py 3617 peak_refine() -     648.9    648.9    0.160
[INFO]    :: edgetrace.py 3617 peak_refine() -     703.6    703.7    1.011
[INFO]    :: edgetrace.py 3617 peak_refine() -     759.9    759.9    0.252
[INFO]    :: edgetrace.py 3617 peak_refine() -     818.2    818.2    0.206
[INFO]    :: edgetrace.py 3617 peak_refine() -     878.8    878.8    0.156
[INFO]    :: edgetrace.py 3617 peak_refine() -     941.7    941.7    0.091
[INFO]    :: edgetrace.py 3617 peak_refine() -    1007.1   1007.2    0.011
[INFO]    :: edgetrace.py 3617 peak_refine() -    1075.3   1075.4    0.043
[INFO]    :: edgetrace.py 3617 peak_refine() -    1146.4   1146.4    0.128
[INFO]    :: edgetrace.py 3617 peak_refine() -    1220.6   1220.7    0.209
[INFO]    :: edgetrace.py 3617 peak_refine() -    1298.3   1298.3    0.299
[INFO]    :: edgetrace.py 3617 peak_refine() -    1465.7   1465.7    0.471
[INFO]    :: edgetrace.py 3617 peak_refine() -    1555.3   1555.3    0.354
[INFO]    :: edgetrace.py 3617 peak_refine() -    1649.6   1649.6    0.227
[INFO]    :: edgetrace.py 3617 peak_refine() -    1749.1   1749.1    0.088
[INFO]    :: edgetrace.py 3617 peak_refine() -    1854.3   1854.4    0.099
[INFO]    :: edgetrace.py 3617 peak_refine() -    1965.9   1965.9    0.289
[INFO]    :: edgetrace.py 3628 peak_refine() - Removing 1 trace(s) due to large RMS difference with previous trace locations.
```